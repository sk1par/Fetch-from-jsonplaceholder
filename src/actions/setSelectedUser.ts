import IUser from "../models/user.interface";

const setSelectedUser= (user?: IUser) => {
    return {
        type: 'SELECTED_USER',
        payload: user
    }
}

export default setSelectedUser;