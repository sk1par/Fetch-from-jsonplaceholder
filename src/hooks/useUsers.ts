import { useEffect, useState } from 'react';
import IUser from '../models/user.interface';
import { checkTimes } from '../utils/utils';
import jsonApi from '../api/jsonApi';
import { LocalStorageUsers } from '../models/localStorageUsers.interface';
import saveToLocalStorage from '../utils/saveToLocalStorage';
import { isTimeExpire } from '../utils/isTimeExpire';

export const useUsers = () => {
    const [users, setUsers] = useState<IUser[]>([]);
    const [error, setError] = useState(false);
    let itemsFromLocalStorage: LocalStorageUsers;


    useEffect(() => {
        itemsFromLocalStorage = JSON.parse(localStorage.getItem('users') || '{}');
        getUsers();
    }, []);

    const getUsers = async () => {
        if (isTimeExpire(itemsFromLocalStorage)) {
            const response = await jsonApi.get('/users');
            if (!response.data.message) {
                setUsers(response.data);
                saveToLocalStorage(response.data);
            } else {
                setError(true);
            }
        } else {
            await setUsers(itemsFromLocalStorage.users);
        }
    };

    return {users, error};
}