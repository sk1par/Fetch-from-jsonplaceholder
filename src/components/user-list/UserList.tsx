import { Link } from 'react-router-dom';
import User from '../user/User';
import IUser from '../../models/user.interface';
import { useDispatch } from 'react-redux';
import { setSelectedUser } from '../../actions';
import {useUsers} from '../../hooks/useUsers';
import ErrorMessage from '../error-message/ErrorMessage';


const UserList = () => {
  const users = useUsers();
  const dispatch = useDispatch();

  const selectUser = (user: IUser) => {
    dispatch(setSelectedUser(user));
  };
  
  return (
    !users.error ? (
      <div className="container" data-auto-id='user-list-div'>
      <div className="col-xs-8">
        {users.users?.map((user: IUser) => (
          <div onClick={() => selectUser(user)} key={user.id} data-auto-id='user-element-div'>
            <Link to={`/user/${user.id}`}>
              <User user={user} />
            </Link>
          </div>
        ))}
      </div>
    </div>
    ) : (<ErrorMessage />)
  )
}

export default UserList
