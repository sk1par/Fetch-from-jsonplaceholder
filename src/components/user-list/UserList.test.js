import { shallow, mount } from 'enzyme';
import UserList from './UserList';
import  setSelectedUser from '../../actions/setSelectedUser';
import { useSelector } from 'react-redux';
import { shallowToJson } from 'enzyme-to-json';
import {useUsers} from '../../hooks/useUsers';

jest.mock('../../actions/setSelectedUser', () => {
    const setSelectedUser = jest.fn();
    return setSelectedUser;
});

jest.mock('../../hooks/useUsers');

const getInitialObject = () => [{
    email: 'test@abv.bg',
    id: 1,
    name: 'test',
    phone: '123132',
    username: 'skipar',
    website: 'abv.bg'
}];

it('should render User User List component correctly', () => {
    const wrapper = shallow(<UserList />);
    const userPreviewDiv = wrapper.find('[data-auto-id="user-list-div"]');
    expect(userPreviewDiv).toHaveLength(1);
    expect(wrapper).toMatchSnapshot();
});

it('should trigger set selected user action correctly after click', () => {
    useUsers.mockReturnValueOnce([getInitialObject()]);
    const wrapper = shallow(<UserList />);
    const item = wrapper.find('[data-auto-id="user-element-div"]');
    item.simulate('click');
    expect(setSelectedUser).toHaveBeenCalledTimes(1);
    expect(setSelectedUser).toHaveBeenCalledWith(getInitialObject());
    expect(shallowToJson(wrapper)).toMatchSnapshot();
});