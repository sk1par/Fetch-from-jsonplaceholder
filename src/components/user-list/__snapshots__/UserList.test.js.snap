// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render User User List component correctly 1`] = `ShallowWrapper {}`;

exports[`should trigger set selected user action correctly after click 1`] = `
<div
  className="container"
  data-auto-id="user-list-div"
>
  <div
    className="col-xs-8"
  >
    <div
      data-auto-id="user-element-div"
      onClick={[Function]}
    >
      <Link
        to="/user/undefined"
      >
        <User
          user={
            Array [
              Object {
                "email": "test@abv.bg",
                "id": 1,
                "name": "test",
                "phone": "123132",
                "username": "skipar",
                "website": "abv.bg",
              },
            ]
          }
        />
      </Link>
    </div>
  </div>
</div>
`;
